### ISSUE TITLE
v0.171.0 breaking change: FullyConnectedLayer<T> dropped (int, int, IActivationFunction<T>?) constructor

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1226]
## Summary

Upgrading from AiDotNet 0.166.0 → 0.171.0 fails to compile downstream consumers due to a removed constructor overload on `FullyConnectedLayer<T>`. The 3-argument `(int inputSize, int outputSize, IActivationFunction<T>? activation)` overload no longer exists.

## Reproduction

In any project consuming AiDotNet 0.166.0 → 0.171.0:

```csharp
using AiDotNet.LinearAlgebra;
using AiDotNet.NeuralNetworks.Layers;
using AiDotNet.ActivationFunctions;

int inputDim = 10, hidden = 16;
var layer = new FullyConnectedLayer<double>(inputDim, hidden, (IActivationFunction<double>?)null);
```

## Expected behavior

Either the overload remains (backward-compat shim) or a clear migration path is documented.

## Actual behavior

Compile error CS1729 across all call sites:
```
error CS1729: 'FullyConnectedLayer<double>' does not contain a constructor that takes 3 arguments
```

## Impact

- HarmonicEngine repository: **12 build errors** across HeadToHead test files (2DMoonsTest, HarmonicMixerHebbianTest, SpectralDiffuser, PerBit, STP, FAvsSGD).
- Forced rollback of the entire AiDotNet package version from 0.171.0 → 0.166.0 to unblock other work.
- Blocks adoption of subsequent fixes shipped in 0.167–0.171, including the Transformer facade fixes referenced in HarmonicEngine issue #79.

## Affected call sites in HarmonicEngine

```
tests/Learning/HeadToHead_DenseLayer_HarmonicMixerHebbianTest.cs:361,362
tests/Learning/HeadToHead_DenseLayer_2DMoonsTest.cs:185,186
tests/Learning/HeadToHead_DenseLayer_2DMoons_STPTest.cs:170,171
tests/Learning/HeadToHead_DenseLayer_SpectralDiffuser_2DMoonsTest.cs:319,320
tests/Learning/HeadToHead_DenseLayer_PerBit_2DMoonsTest.cs:192,193
tests/Learning/HeadToHead_FAvsSGD_LowRungMatchedParam.cs:435,436
```

All 6 files use the same 3-argument call pattern: `new FullyConnectedLayer<T>(inputSize, outputSize, (IActivationFunction<T>?)null)`.

## Requested resolution

Pick one:
1. **Restore the `(int, int, IActivationFunction<T>?)` constructor overload** (preferred — backward-compat) so existing call sites compile without changes.
2. **Document the new constructor signature** in CHANGELOG / migration guide for v0.171.0 so consumers can do a targeted call-site update.

If option 2 is preferred, please publish:
- The current required constructor signature
- Whether `IActivationFunction<T>?` is now passed differently (named arg? property? builder pattern?)
- Whether biases/initialization are now mandatory or moved elsewhere

## Repro environment

- AiDotNet versions: 0.166.0 (works) → 0.171.0 (fails)
- .NET 10.0
- Windows 11 / VS Code
- HarmonicEngine commit `5375705` (rollback to 0.166.0 to unblock)

## Related

HarmonicEngine issue #79 — Phase A4 Transformer baseline depends on adopting the Transformer facade fixes shipped in 0.167–0.171. Currently blocked by this constructor break preventing the upgrade.

**Comment by ooples** [comment]
Closing — this is intentional API migration, not a bug.

## Why the 3-arg constructor was removed

PR #1218 (commit `4b28b540d`, *"refactor(layers): migrate FullyConnectedLayer to lazy shape inference"*) intentionally removed the 3-arg `(inputSize, outputSize, activation)` constructor as part of the broader PyTorch-style lazy-shape-inference migration covering ~50 layers. The decision was **lazy-by-default**: input feature size is resolved from `input.Shape[^1]` on first forward, so callers don't pin it at construction.

This is a stronger design choice than PyTorch's, which keeps `nn.Linear(in, out)` and `nn.LazyLinear(out)` as separate classes. We chose to make `FullyConnectedLayer` lazy-only.

## Migration recipe (HarmonicEngine)

Drop the leading `inputSize` argument at every call site. Mechanical 1-line change per site:

```csharp
// Before (v0.166.0 and earlier)
new FullyConnectedLayer<T>(inputSize, outputSize, (IActivationFunction<T>?)null)

// After (v0.171.0+)
new FullyConnectedLayer<T>(outputSize, (IActivationFunction<T>?)null)
```

Applied to the 6 files in your list:

```
tests/Learning/HeadToHead_DenseLayer_HarmonicMixerHebbianTest.cs:361,362
tests/Learning/HeadToHead_DenseLayer_2DMoonsTest.cs:185,186
tests/Learning/HeadToHead_DenseLayer_2DMoons_STPTest.cs:170,171
tests/Learning/HeadToHead_DenseLayer_SpectralDiffuser_2DMoonsTest.cs:319,320
tests/Learning/HeadToHead_DenseLayer_PerBit_2DMoonsTest.cs:192,193
tests/Learning/HeadToHead_FAvsSGD_LowRungMatchedParam.cs:435,436
```

Each line just needs the first int dropped from the constructor call.

## What weights look like now

Same `_weights [outputSize, inputSize]` and `_biases [outputSize]` as before — they're allocated on the first `Forward(input)` call rather than in the constructor. `ParameterCount` returns 0 until that first forward; downstream code that reads parameters before any forward needs to either run a warm-up forward or call `EnsureInitialized` (also added by #1218).

## Other layers covered by the same migration

The same 1-line drop-the-input-size pattern applies to `ConvolutionalLayer`, `LSTMLayer`, `GRULayer`, `RecurrentLayer`, the Transformer Encoder/Decoder layers, `BidirectionalLayer`, `TimeDistributedLayer`, plus the spatial/feature-rigid layers tracked in #1209. If you hit additional CS1729s on those, they all migrate the same way.

**Comment by ooples** [comment]
Closing my own report — this was a misdiagnosis on my end. AiDotNet 0.171.0 adopted PyTorch-style lazy layers where input shape is inferred at first forward pass, so removing the eager (int inputSize, int outputSize, IActivationFunction<T>?) constructor was the **correct API improvement**, not a regression. Downstream consumers (HarmonicEngine) need to migrate to the lazy-layers API: drop inputSize from construction, let the layer infer shape from the first forward call. Apologies for the noise.
